device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/exceptions/MultiplexingClientDeviceRegistrationTimeoutException.java

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.sdk.iot.device.exceptions;
+
+/**
+ * Exception that is thrown when a multiplexed device client registration or unregistration times out.
+ */
+public class MultiplexingClientDeviceRegistrationTimeoutException extends MultiplexingClientException
+{
+    /**
+     * Construct a new MultiplexingClientDeviceRegistrationTimeoutException with no nested exception and no error message.
+     */
+    public MultiplexingClientDeviceRegistrationTimeoutException()
+    {
+        super();
+    }
+
+    /**
+     * Construct a new MultiplexingClientDeviceRegistrationTimeoutException with no nested exception but with an error message.
+     * @param message The top level message for this exception.
+     */
+    public MultiplexingClientDeviceRegistrationTimeoutException(String message)
+    {
+        super(message);
+    }
+
+    /**
+     * Construct a new MultiplexingClientDeviceRegistrationTimeoutException with a nested exception and an error message.
+     * @param message The top level message for this exception.
+     * @param cause The nested exception.
+     */
+    public MultiplexingClientDeviceRegistrationTimeoutException(String message, Throwable cause)
```

Comment:
q - how detailed is this message/ throwable? does it give any info about which device actually timed out (while opening/ closing)?

Reply:
The message will be a hardcoded "Timed out waiting for all device unregistrations to finish." and there will not be a nested throwable, at least for now. The only reason I provide overloaded constructors that take throwables is because that is standard for exceptions.

> does it give any info about which device actually timed out 

No, because this timeout is for all of the device registrations, not per device registration. If a user calls the singular multiplexingClient.registerDeviceClient(), then they will know which client timed out. For the bulk APIs, though, I expect users will just call the API again with the same set of device clients to be registered. The multiplexing client allows for you to "re-register" a client or a set of clients without throwing, after all.